src/dbnode/storage/index_queue_forward_write_test.go

```diff
@@ -0,0 +1,565 @@
+	xtime "github.com/m3db/m3/src/x/time"
+
+	"github.com/fortytw2/leaktest"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/uber-go/tally"
+)
+
+func generateOptionsNowAndBlockSize() (Options, time.Time, time.Duration) {
+	idxOpts := testNamespaceIndexOptions().
+		SetInsertMode(index.InsertSync).
+		SetForwardIndexProbability(1).
+		SetForwardIndexThreshold(1)
+
+	opts := DefaultTestOptions().
+		SetIndexOptions(idxOpts)
+
+	var (
+		retOpts        = opts.SeriesOptions().RetentionOptions()
+		blockSize      = retOpts.BlockSize()
+		bufferFuture   = retOpts.BufferFuture()
+		bufferFragment = blockSize - time.Duration(float64(bufferFuture)*0.5)
+		now            = time.Now().Truncate(blockSize).Add(bufferFragment)
+
+		clockOptions = opts.ClockOptions()
+	)
+
+	clockOptions = clockOptions.SetNowFn(func() time.Time { return now })
+	opts = opts.SetClockOptions(clockOptions)
+
+	return opts, now, blockSize
+}
+
+func setupForwardIndex(
+	t *testing.T,
+	ctrl *gomock.Controller,
+) (namespaceIndex, time.Time, time.Duration) {
+	newFn := func(
+		fn nsIndexInsertBatchFn,
+		md namespace.Metadata,
+		nowFn clock.NowFn,
+		s tally.Scope,
+	) namespaceIndexInsertQueue {
+		q := newNamespaceIndexInsertQueue(fn, md, nowFn, s)
+		q.(*nsIndexInsertQueue).indexBatchBackoff = 10 * time.Millisecond
+		return q
+	}
+
+	md, err := namespace.NewMetadata(defaultTestNs1ID, defaultTestNs1Opts)
+	require.NoError(t, err)
+
+	opts, now, blockSize := generateOptionsNowAndBlockSize()
+	idx, err := newNamespaceIndexWithInsertQueueFn(md, newFn, opts)
+	assert.NoError(t, err)
+
+	var (
+		ts   = idx.(*nsIndex).state.latestBlock.StartTime()
+		next = ts.Truncate(blockSize).Add(blockSize)
+		id   = ident.StringID("foo")
+		tags = ident.NewTags(
+			ident.StringTag("name", "value"),
+		)
+		lifecycle = index.NewMockOnIndexSeries(ctrl)
+	)
+
+	lifecycle.EXPECT().OnIndexFinalize(xtime.ToUnixNano(ts))
+	lifecycle.EXPECT().OnIndexSuccess(xtime.ToUnixNano(ts))
+
+	lifecycle.EXPECT().OnIndexFinalize(xtime.ToUnixNano(ts.Add(blockSize)))
+	lifecycle.EXPECT().OnIndexSuccess(xtime.ToUnixNano(ts.Add(blockSize)))
+
+	lifecycle.EXPECT().NeedsIndexUpdate(xtime.ToUnixNano(next)).Return(true)
+	lifecycle.EXPECT().OnIndexPrepare()
+
+	entry, doc := testWriteBatchEntry(id, tags, now, lifecycle)
+	batch := testWriteBatch(entry, doc, testWriteBatchBlockSizeOption(blockSize))
+	assert.NoError(t, idx.WriteBatch(batch))
+
+	return idx, now, blockSize
+}
+
+func TestNamespaceForwardIndexInsertQuery(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	defer leaktest.CheckTimeout(t, 2*time.Second)()
+
+	ctx := context.NewContext()
+	idx, now, blockSize := setupForwardIndex(t, ctrl)
+	defer idx.Close()
+
+	reQuery, err := m3ninxidx.NewRegexpQuery([]byte("name"), []byte("val.*"))
+	assert.NoError(t, err)
+
+	// NB: query both the current and the next index block to ensure that the
```

Comment:
I don't think this is confirming that the write exists in the next block -- cause you're querying both current and next block. Need to ensure you're checking both individually, no?

Reply:
It should be fine, `now` is 5 minutes before the block boundary, and this will check the range which is from 6 to 4 minutes before the boundary to check for the current block, and then 6 to 4 minutes before the next block boundary to check only the future block